### PR TITLE
feat: CODECOPY opcode

### LINKED ISSUE #39: Opcode: CODECOPY

### PR DESCRIPTION
**Pull request by ricomateo**
Closes #39 
This PR implements code generation for [CODECOPY](https://www.evm.codes/#39?fork=cancun) opcode.

### CODE REVIEW

**Review by FrancoGiachetta — APPROVED** [review]
LGTM

File: src/main.rs [hunk 1]
```diff
@@ -1,24 +1,15 @@
...
     program::Program,
-    syscall::SyscallContext,
+    Evm,
 };
+use num_bigint::BigUint;
 
 fn main() {
     let args: Vec<String> = std::env::args().collect();
     let path = args.get(1).expect("No path provided").as_str();
-    let opt_level = match args.get(2).map(String::as_str) {
```
Comment (MegaRedHand) [inline src/main.rs:15]: Please revert changes to this file.

File: src/program.rs [hunk 2]
```diff
@@ -296,6 +296,9 @@ impl TryFrom<u8> for Opcode {
             x if x == Opcode::LOG3 as u8 => Opcode::LOG3,
             x if x == Opcode::LOG4 as u8 => Opcode::LOG4,
             x if x == Opcode::RETURN as u8 => Opcode::RETURN,
+            x if x == Opcode::MSTORE as u8 => Opcode::MSTORE,
+            x if x == Opcode::MSTORE8 as u8 => Opcode::MSTORE8,
```
Comment (MegaRedHand) [inline src/program.rs:300]: These where duplicated

File: src/syscall.rs [hunk 3]
```diff
@@ -130,6 +130,8 @@ pub struct InnerContext {
 #[derive(Debug)]
 pub struct SyscallContext<'c> {
     pub env: Env,
+    // The program bytecode
+    pub program: Vec<u8>,
```
Comment (MegaRedHand) [inline src/syscall.rs:134]: This should go inside the `inner_context`

File: src/syscall.rs [hunk 4]
```diff
@@ -248,6 +251,28 @@ impl<'c> SyscallContext<'c> {
...
+        // adjust the size so it does not go out of bounds
+        let size: u32 = if code_offset + size > code_size {
+            code_size - code_offset
+        } else {
+            size
+        };
+        // copy the program into memory
+        for (i, j) in (code_offset..code_offset + size).enumerate() {
+            self.inner_context.memory[dest_offset as usize + i] = self.program[j as usize];
+        }
```
Comment (MegaRedHand) [inline src/syscall.rs:274]: We should use [`Vec::copy_from_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.copy_from_slice) here.

File: src/syscall.rs [hunk 5]
```diff
@@ -248,6 +251,28 @@ impl<'c> SyscallContext<'c> {
...
+        dest_offset: u32,
+    ) {
+        let code_size = self.program.len() as u32;
+        // if the offset is out of bounds then nothing is copied
+        if code_offset >= code_size {
+            return;
+        }
+        // adjust the size so it does not go out of bounds
+        let size: u32 = if code_offset + size > code_size {
+            code_size - code_offset
```
Comment (MegaRedHand) [inline src/syscall.rs:267]: You can do:

```rust
code_size.checked_sub(code_offset).unwrap_or(0)
```

File: src/constants.rs [hunk 6]
```diff
@@ -69,6 +71,11 @@ pub mod gas_cost {
         (new_memory_cost - last_memory_cost).into()
     }
 
+    pub fn codecopy_dynamic_gas_cost(size: u32) -> i64 {
+        let minimum_word_size = (size + 31) / 32;
+        (minimum_word_size * 3) as i64
+    }
```
Comment (MegaRedHand) [inline src/constants.rs:77]: This looks the same as `memory_copy_cost`. Can we reuse that?

File: src/utils.rs [hunk 7]
```diff
@@ -1020,6 +1020,69 @@ pub(crate) fn compute_memory_cost<'c>(
     Ok(memory_cost)
 }
 
+pub(crate) fn compute_codecopy_gas_cost<'c>(
```
Comment (MegaRedHand) [inline src/utils.rs:1023]: Isn't this the same as `compute_copy_cost`?

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: src/syscall.rs (hunk 5 above)
Comment (ricomateo) [reply src/syscall.rs:267]: I'm trying to do it that way but the linter is throwing this warning 
`help: consider using saturating_sub: code_size.saturating_sub(code_offset)`
 however `saturating_sub` is not useful in this case

Comment (MegaRedHand) [reply src/syscall.rs:272]: `saturating_sub` is _just_ what we need

File: src/syscall.rs [hunk 8]
```diff
@@ -121,6 +121,8 @@ pub struct InnerContext {
     memory: Vec<u8>,
     /// The result of the execution
     return_data: Option<(usize, usize)>,
+    // The program bytecode
+    pub program: Vec<u8>,
```
Comment (JulianVentura) [inline src/syscall.rs:125]: I have a question. Why do we need to store the program here? Can't we access the `Db` to get the bytecode for the callee address?
We access the `TransactTo` address like we do in `ADDRESS` opcode.
Then with that address we access the `Db` to get the bytecode, like we do in `EXTCODESIZE`.

File: tests/evm.rs [hunk 9]
```diff
@@ -696,6 +696,78 @@ fn log4() {
...
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Codecopy,
+        Operation::Push((1_u8, BigUint::from(size))),
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Return,
+    ]
+    .into();
+
+    let mut env = Env::default();
+    env.tx.gas_limit = 999_999;
```
Comment (MegaRedHand) [inline tests/evm.rs:716]: This shouldn't be needed

File: tests/evm.rs [hunk 10]
```diff
@@ -696,6 +696,78 @@ fn log4() {
...
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Codecopy, // 6th byte
+        Operation::Push((1_u8, BigUint::from(size))),
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Return,
+    ]
+    .into();
+
+    let mut env = Env::default();
+    env.tx.gas_limit = 999_999;
```
Comment (MegaRedHand) [inline tests/evm.rs:753]: Neither this one

**Review by MegaRedHand — APPROVED** [review]
LGTM

File: src/syscall.rs (hunk 8 above)
Comment (MegaRedHand) [reply src/syscall.rs:125]: Let's leave this for another PR

File: tests/evm.rs [hunk 11]
```diff
@@ -696,6 +696,77 @@ fn log4() {
...
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Codecopy,
+        Operation::Push((1_u8, BigUint::from(size))),
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Return,
+    ]
+    .into();
+
+    let mut env = Env::default();
+    dbg!(env.tx.gas_limit);
```
Comment (MegaRedHand) [inline tests/evm.rs:716]:
```suggestion
```

**Pull request merged**